warcio/recordloader.py

```diff
@@ -126,22 +121,46 @@ def parse_record_stream(self, stream,
         # limit stream to the length for all valid records
         if length is not None and length >= 0:
             stream = LimitReader.wrap_stream(stream, length)
-
+            stream, is_verifying = self.wrap_digest_verifying_stream(stream, rec_type,
+                                                                     rec_headers,
+                                                                     length=length,
+                                                                     check_digests=check_digests)
+        else:
+            is_verifying = False
 
         http_headers = None
 
         # load http headers if parsing
         if not no_record_parse:
             http_headers = self.load_http_headers(rec_type, uri, stream, length)
+        else:
+            is_verifying = False  # can't find the start of the payload
+            # XXX resource record might have no headers, a payload digest, and no block digest?
 
         # generate validate http headers (eg. for replay)
         if not http_headers and ensure_http_headers:
             http_headers = self.default_http_headers(length, content_type)
 
+        if is_verifying:
+            stream.begin_payload()
+
         return ArcWarcRecord(the_format, rec_type,
                              rec_headers, stream, http_headers,
                              content_type, length)
 
+    def wrap_digest_verifying_stream(self, stream, rec_type, rec_headers, length=None, check_digests=False):
+        payload_digest = rec_headers.get_header('WARC-Payload-Digest')
```

Comment:
```suggestion
        if not check_digests:
            return stream, False 
        payload_digest = rec_headers.get_header('WARC-Payload-Digest')
```
I think it would be wise to just return the original stream when `check_digests` is false because not all WARC generators add `WARC-*-Digest` fields and can avoid wrapping the original stream twice in this case.

Reply:
check_digests=False currently prints a warning if the digest fails.

That's one of the high-level questions that needs other opinions; I worry that people will never see bad digests if the default is no check at all. However, if it turns out that lots of old warcs have bad digests, then this warning will be ugly & would need suppression, which would be ugly code.

Reply:
I avoided wrapping in the case where there is no payload or block digest at all.